jooq-pg/src/main/resources/db/migration/ehr/V15__data_tables_add_parent_num.sql

```diff
@@ -16,137 +16,275 @@
+                r.num_cap = r.num;
+                h[r.entity_idx_len] = r.num;
+            ELSEIF  r.entity_idx_len < prev_r.entity_idx_len THEN
+                --parent
+                r.num_cap = h[r.entity_idx_len];
+            ELSE -- r.entity_idx_len > prev_r.entity_idx_len
+            --different ancestor
+                r.num_cap = r.num;
+                for j in prev_r.entity_idx_len .. r.entity_idx_len LOOP
+                        h[j] = r.num;
+                    END LOOP;
+            END IF;
+            rs[i] = r;
+            prev_r = r;
+        END LOOP;
+    return rs;
+END;$function$;
+
+--
+-- Select structures from affected table
+--
+CREATE OR REPLACE FUNCTION ext.mig_retrieve_nums_batch(
+    table_name text,
+    id_exp text,
+    ehr_id_exp text,
+    folders_idx_exp text,
+    version_exp text,
+    mig_idx text,
+    batch_size integer
+) RETURNS ext.mig_num_type[]
+    LANGUAGE plpgsql
+    IMMUTABLE STRICT
+AS $function$
+DECLARE
+    rs ext.mig_num_type[];
+BEGIN
+    EXECUTE format('
+        SELECT array_agg(sub.r)
+        FROM(
+            SELECT row(
+                VO_ID,
+                NUM,
+                %5$s,           --EHR_ID
+                ENTITY_IDX_LEN,
+                %3$s,           --SYS_VERSION
+                0,              --parent_num
+                -1              --num_cap
+            ) as r
+            FROM (
+               SELECT %2$s AS id
+               FROM %4$s v
+               WHERE num = 0 and num_cap = -1 and %3$s = 1
+               LIMIT $1
+            ) ids
+            JOIN %4$s d ON ids.id = d.%2$s
+            ORDER BY ids.id, %3$s, d.num
+        ) sub', mig_idx, id_exp, version_exp, table_name, ehr_id_exp, folders_idx_exp)
+        INTO rs
+        USING batch_size;
+    return rs;
+END;$function$;
+
+--
+-- Migration procedure
+--
+CREATE OR REPLACE PROCEDURE ext.mig_num_columns(
+    rel_name regclass,
+    id_exp text,
+    mig_idx text,
+    batch_size integer DEFAULT 1000
+) LANGUAGE plpgsql
+AS $procedure$
+DECLARE
+    rs ext.mig_num_type[];
+    cols text;
+    ehr_id_exp text;
+    folders_idx_exp text;
+    version_exp text;
+    updated_cnt integer := -1;
+    err_msg text;
+    err_state text;
+    last timestamp;
+BEGIN
+
+    RAISE NOTICE 'Starting migration for %', rel_name;
+
+    SELECT string_agg(attname, ',')
+    FROM (
+             SELECT attname
+             FROM pg_attribute
+             WHERE attrelid=rel_name AND NOT attisdropped AND attnum > 0
+             ORDER BY attnum
+         ) att
+    INTO cols;
+
+    ehr_id_exp=COALESCE(substring(cols FROM '(?i)ehr_id'),'NULL');
+    folders_idx_exp=COALESCE(substring(cols FROM '(?i)ehr_folders_idx'),'NULL');
+    version_exp=COALESCE(substring(cols FROM '(?i)sys_version'),'1::int');
+
+    last = clock_timestamp();
+
+    WHILE updated_cnt <> 0 LOOP
+            COMMIT;
+            SET TRANSACTION ISOLATION LEVEL SERIALIZABLE READ ONLY DEFERRABLE;
```

Comment:
Separate transactions are only needed for pre-flight.
The initial and intermediate commit are not needed. 
Also, catching exceptions is not needed.

Reply:
Intension was to be able to use this script as a pre-flight.

Reply:
Not sure about the impact, but it needs 3x as many transactions + transaction isolation

Reply:
it's removed in this commit -> https://github.com/ehrbase/ehrbase/pull/1389/commits/781dae2cc444cf78bd0c7e43ea7daa433b6bf6a6